Formulation refactoring to remove ee_decision variables and fix jittering in the wheels motions.

viviansuzano opened the pull request:
The main changes are:
- Refactor of the formulation and the phase-based parametrization for the wheels' motions, removing the ee_decision variables (that were not being optimized over). 
- Add a few more constraints to ensure consistency with wheeled locomotion, including the constraint that prevents jittering and spikes in the wheels' acceleration.
- Add some new parameters that help improving the motions.
- Remove unused terrain types, focusing on discontinuous obstacles.
- Add a configuration file for easier change of some parameters. Feel free to add more parameters to the file, if you see fit.
- I've kept two previous methods called "AdvancedInititialisationBase" and "AdvancedInititialisationEE", bt they are not used and can be removed, if necessary.
- Since we need a specific gait to handle some terrains, I have limited the gait options for each terrain in the user interface, preventing an eventual non-convergence of the solver. For testing with all possible gaits, just comment a few lines on the file towr_user_interface.cc

mbjelonic committed:
Added bag file of elevation map.

mbjelonic committed:
Changed step.

mbjelonic committed:
Increased size of grid map.

mbjelonic committed:
Reduced size of grids.

Commit:
- Refactor of the formulation and the phase-based parametrization for the wheels' motions, removing the ee_decision variables (that were not being optimized over).
- Add a few more constraints to ensure consistency with wheeled locomotion, including the constraint that prevents jittering and spikes in the wheels' acceleration.
- Add some new parameters that help improving the motions.
- Remove unused terrain types, focusing on discontinuous obstacles.
- Add a configuration file for easier change of some parameters. Feel free to add more parameters to the file, if you see fit.
- Since we need a specific gait to handle some terrains, I have limited the gait options for each terrain in the user interface, preventing an eventual non-convergence of the solver. For testing with all possible gaits, just comment a few lines on the file towr_user_interface.cc

Commit:
Remove some comments.

rubengrandia commented on towr/include/towr/constraints/base_acc_limits_constraint.h:
```diff
@@ -0,0 +1,47 @@
+namespace towr {
+
+// Limits the acceleration of the base
+class BaseAccLimitsConstraint : public TimeDiscretizationConstraint {
```
Needs Eigen_Aligned_Allocator_New

rubengrandia replied:
Same with other classes below, when they have a Vector3d as member

rubengrandia commented on towr/include/towr/constraints/ee_acc_limits_constraint.h:
```diff
@@ -0,0 +1,49 @@
+  using EE = uint;
+  using Vector3d = Eigen::Vector3d;
+
+  EEAccLimitsConstraint (Vector3d acc_max, const EE& ee,
```
Would take acc_max by const&, ee by value

rubengrandia commented on towr/include/towr/constraints/wheels_lateral_constraint.h:
```diff
@@ -0,0 +1,67 @@
+
+  std::vector<double> GetVectorGlobalTimeAtNodes (VecDurations poly);
+  Eigen::Matrix3d GetJacobianTerrainBasis(HeightMap::Direction basis, double x, double y) const;
+  Jacobian SkewSymmetricMatrix (const Vector3d v) const;
```
const &

rubengrandia reviewed (commented):
Lots of work! Thanks for the effort @viviansuzano .
It is a bit much to go through in detail, but it would be great if you can check some things related to Eigen to avoid alignement bugs.
1) If classes have fixed-size member, please add the aligned allocator new.
2) Don't pass fixed-size eigen objects by value. this can also cause alignement issues.
https://eigen.tuxfamily.org/dox/group__TopicPassingByValue.html

pajovic23 committed:
Smart initialization added. New class StepsGaitGenerator added. New terrain Steps added. Ruben's
comments added for eigen class inside the other class.
Remarks for the Steps terrain:
 - The solution can vary based on the desired speed of the robot. This could be changed by
   changing the goal, or by changing the time of the simulation.
 - There are two parameters: step_margins and swing_phase_duration, which could be also be tuned
   in order for solver to find the solution, but this cannot be changed in the runtime
 - Since the StepsGaitGenerator is used, the towr will break, if the generated are less or equal
   zero. This could happen if: - step_margins or swing_phase_durations are too high.
                               - the goal is in the middle of the terrain. (that is why the goal
                                 the goal should be always set at the end of the last step +
                                 some small margin in order for robot to finish entirely on the
                                 last step)

mbjelonic committed:
Removed some unnecessary launch file arguements.

mbjelonic committed:
Removed some unnecessary launch file arguements.

mbjelonic committed:
Changed the amount of nodes.

mbjelonic committed:
Changed the amount of nodes.

pajovic23 committed:
Fixed implementation of smart initialization and speed reduction of the base added around steps.

pajovic23 committed:
merge from pull

pajovic23 committed:
terrain for tests added as comments in Steps

mbjelonic committed:
Minor change.

mbjelonic committed:
Merge pull request #8 from leggedrobotics/dev/master_wheels_initialization

Smart initialization added. New class StepsGaitGenerator added. New t…
